lib/EarlyFunctionAnalysis/ApplyRegisterStateDeductions.cpp

```diff
@@ -0,0 +1,188 @@
+    bool SecondYesOrDead = isYesOrDead(State.at(Second).AsAnReturnValue);
+    if (FirstYesOrDead && SecondYesOrDead) {
+      // Both `First` and `Second` shouldn't be positive.
+      if constexpr (EnsureABIConformance) {
+        // There's no way to determine which one should be nullified. Abort.
+        // TODO: maybe we should notify the caller about what happened.
+        return std::nullopt;
+      } else {
+        return std::nullopt;
+      }
+    }
+  }
+
+  return State; // Return a copy of the original state if everything is fine.
+}
+
+template<typename CallingConvention, bool EnsureABIConformance = false>
+std::optional<RegisterStateMap>
+applyNonPositionBasedDeductions(const RegisterStateMap &State) {
+  if constexpr (CallingConvention::OnlyStartDoubleArgumentsFromAnEvenRegister) {
+    // Only partial deduction is possible. It can be implemented if necessary.
+    return std::nullopt;
+  }
+
+  auto ArgumentAccessor = [](auto &FS) -> auto & { return FS.AsAnArgument; };
+  auto RVAccessor = [](auto &FS) -> auto & { return FS.AsAnReturnValue; };
+
+  auto &GPAR = CallingConvention::GeneralPurposeArgumentRegisters;
+  auto &VAR = CallingConvention::VectorArgumentRegisters;
+  auto &GPRVR = CallingConvention::GeneralPurposeReturnValueRegisters;
+  auto &VRVR = CallingConvention::VectorReturnValueRegisters;
+
+  auto RelevantGPAR = getRegisterSubrange(GPAR, State, ArgumentAccessor);
+  auto RelevantVAR = getRegisterSubrange(VAR, State, ArgumentAccessor);
+  auto RelevantGPRVR = getRegisterSubrange(GPRVR, State, ArgumentAccessor);
+  auto RelevantVRVR = getRegisterSubrange(VRVR, State, ArgumentAccessor);
+
+  using CRegister = const model::Register::Values;
+  auto ArgumentRegisters = llvm::concat<CRegister>(RelevantGPAR, RelevantVAR);
+  auto RetValRegisters = llvm::concat<CRegister>(RelevantGPRVR, RelevantVRVR);
+
+  RegisterStateMap Result = State; // Copy the state.
+
+  for (auto &[Register, RegisterState] : Result) {
+    auto &[AsAnArgument, AsAnReturnValue] = RegisterState;
+
+    if (AsAnArgument == model::RegisterState::Invalid
+        || AsAnArgument == model::RegisterState::Contradiction) {
+      // The input is not valid - abort the deduction.
+      return std::nullopt;
+    }
+
+    auto ArgumentIterator = llvm::find(ArgumentRegisters, Register);
+    bool IsArgumentRequired = ArgumentIterator != ArgumentRegisters.end();
+    if (model::RegisterState::isYesOrDead(AsAnArgument)) {
+      if (IsArgumentRequired) {
+        // Current register is already marked "Yes" or "Dead", do nothing.
+      } else {
+        // Current register is marked "Yes" or "Dead" but it shouldn't be.
+        if constexpr (EnsureABIConformance)
+          AsAnArgument = model::RegisterState::No;
+        else
+          return std::nullopt;
+      }
+    } else {
+      if (IsArgumentRequired) {
+        // Current register is not marked as "Yes" or "Dead" but it should be.
+        AsAnArgument = model::RegisterState::YesOrDead;
```

Comment:
That's the way it was done in the example implementation I started with. If the argument is not `Yes` or `Dead` or `YesOrDead` already - make is so.

Reply:
Mmmmh, are you sure?
I think we were forcing to yes all the registers associated to an argument index lower than the highest indexed argument register.

`rsi` is the second argument. If it's yes, and `rdi` (the first) is Maybe, we want to go to Yes.
Here you're forcing all of the arguments to be Yes, which is too aggressive.

Reply:
To avoid any possible confusion, here's how the current algorithm works:

- use `getRegisterSubrange` to get all the registers before "the last used ones".
- concatenate all the resulting subranges (so we have containers with all the registers that are "Required".
- Then for each register in the input State:
  - abort if it's `invalid` or `contradiction` (TODO)
  - if it's `Yes` or `Dead` or `YesOrDead`
    - if it's "Required" do nothing, otherwise set it to `No`.
  - otherwise
    - if it's "Required" set it to `YesOrDead`, otherwise do nothing.

Reply:
Ah, sorry, my brain ignored this part:

```c++

  auto RelevantGPAR = getRegisterSubrange(GPAR, State, ArgumentAccessor);
  auto RelevantVAR = getRegisterSubrange(VAR, State, ArgumentAccessor);
  auto RelevantGPRVR = getRegisterSubrange(GPRVR, State, ArgumentAccessor);
  auto RelevantVRVR = getRegisterSubrange(VRVR, State, ArgumentAccessor);

  using CRegister = const model::Register::Values;
  auto ArgumentRegisters = llvm::concat<CRegister>(RelevantGPAR, RelevantVAR);
  auto RetValRegisters = llvm::concat<CRegister>(RelevantGPRVR, RelevantVRVR);
```

I was thinking it was just some renaming as above:

```c++
  auto &GPAR = CallingConvention::GeneralPurposeArgumentRegisters;
  auto &VAR = CallingConvention::VectorArgumentRegisters;
  auto &GPRVR = CallingConvention::GeneralPurposeReturnValueRegisters;
  auto &VRVR = CallingConvention::VectorReturnValueRegisters;
```

Gotcha.